planner,expression,util: scatter hotspot index in certain scenes by expression index

TonsnakeLin opened the pull request:
Signed-off-by: TonsnakeLin <lpbgytong@163.com>

<!--

Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?
Issue Number: close #31040

It propose a solution to solve the problem that hot indexes hinder the scalability of TiDB in certain case. 
RFC: [84](https://github.com/tikv/rfcs/pull/84)

### What is changed and how it works?

### Check List

Tests <!-- At least one of them must be included. -->

- [ ] Unit test
- [ ] Integration test


Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains experimental features

### Release note

<!-- bugfix or new feature needs a release note -->

```release-note
None
```

cfzjywxk commented on planner/core/rule_predicate_push_down.go:
```diff
@@ -118,6 +127,7 @@ func (p *LogicalUnionScan) PredicatePushDown(predicates []expression.Expression)
 func (ds *DataSource) PredicatePushDown(predicates []expression.Expression) ([]expression.Expression, LogicalPlan) {
 	predicates = expression.PropagateConstant(ds.ctx, predicates)
 	predicates = DeleteTrueExprs(ds, predicates)
+	predicates = ds.AddPrefix4ShardIndexes(ds.ctx, predicates)
```
Should we do the transformation processing push donw conditions or doing the expression rewrite or generation(https://github.com/pingcap/tidb/blob/master/planner/core/expression_rewriter.go#L1031)?
I'm not sure if the expression index filter range could be calculated automatically by now.

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,380 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+			 []*valueInfo              the values of every columns in the returned new condtions
+			 bool                      true means "IN" function is transformed to a "OR" function
+*/
+func AddGcColumnCond(sctx sessionctx.Context,
```
`The inside used function may produce errors do we need to return error in the results?`, yes, there are two ways to process the abnormal cases. first one, returns errors. second, ignore the error an don't add tidb_shard prefix leaving the condition as original

TonsnakeLin replied:
Ok, I'll fix it

TonsnakeLin commented on planner/core/rule_predicate_push_down.go:
```diff
@@ -638,3 +648,108 @@ func (p *LogicalMemTable) PredicatePushDown(predicates []expression.Expression)
+	}
+
+	newConds := make([]expression.Expression, 0, len(conds))
+	newConds = append(newConds, conds...)
```
may be optimize for some cases

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+// (1) columns in accessCond are all columns of the index except the first.
+// (2) the first param of "IN" function shoulde be a column not a expression like `a + b`
+// (3) the rest params of "IN" function all shoulde be constant
+// (4) the first param of "IN" function shoulde be the column in the expression of first index field.
```
first

TonsnakeLin replied:
Yes, It means the expression is the first field of shard index

zhangjinpeng1987 commented:
@TonsnakeLin Is it possible to split this PR into 2 or 3 smaller PRs?

you06 commented on planner/core/rule_column_pruning.go:
```diff
@@ -282,7 +282,9 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column) error {
-		if !used[i] && !exprUsed[i] {
+		// if ds has a shard index, it can't prune the generated column of shard index
+		// so add the new condtion `!ds.containExprPrefixUk`
+		if !used[i] && !exprUsed[i] && !ds.containExprPrefixUk {
```
Does it means that if there is a shard index, no columns will be pruned?

TonsnakeLin replied:
yes, it will prune the tidb_shard() gc column!

TonsnakeLin commented:
> @TonsnakeLin Is it possible to split this PR into 2 or 3 smaller PRs?
All code of this PR has 1157 lines code, and there are about 550 lines test code. I can split it into a PR that supports integer shard index and a PR testing integer shard index.
@zhangjinpeng1987 @cfzjywxk

cfzjywxk commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+		return conditions, nil
+	}
+
+	// remove the accesses from newConditions
```
Better to comment about the meaning of `accesses` and `newAccesses`.

TonsnakeLin replied:
ok

you06 commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+		return true
+	}
+
+	return false
```
What about `return f.FuncName.L == ast.EQ || f.FuncName.L == ast.In`.

TonsnakeLin replied:
oh ... in my opinion，the two ways are both good

you06 commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+}
+
+// IsEqOrInFunc check whetere it si equal or in function
+func IsEqOrInFunc(expr expression.Expression) bool {
```
Is this function only used for testing? If so, we may move it into the test package.

TonsnakeLin replied:
No, it is used in the normal code

you06 commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+		return false
+	}
+
+	return true
```
What about `return ok && col.Equal(nil, cols[1])`

you06 reviewed (commented):
I also noticed that there are many typos in the comments.

you06 commented on expression/builtin_miscellaneous_test.go:
```diff
@@ -591,3 +591,47 @@ func TestBinToUUID(t *testing.T) {
 }
+
+func TestTidbShard(t *testing.T) {
+	t.Parallel()
```
Remove the parallel

TonsnakeLin commented on expression/builtin_miscellaneous.go:
```diff
@@ -1304,3 +1306,49 @@ func swapStringUUID(str string) string {
+	if hashed, err = vitess.HashUint64(uint64(shardKeyInt)); err != nil {
+		return 0, true, err
+	}
+	hashed = hashed % 256
```
yes, I will fix it

TonsnakeLin commented on expression/column.go:
```diff
@@ -680,3 +681,37 @@ func SortColumns(cols []*Column) []*Column {
+// IndexColToExpressionCol return the expression.Column by model.IndexColumn
+func IndexColToExpressionCol(colInfos []*model.ColumnInfo, cols []*Column,
+	col *model.IndexColumn) *Column {
+	return indexCol2Col(colInfos, cols, col)
```
Yes, I wanted to do it at first.

TonsnakeLin commented on planner/core/logical_plan_builder.go:
```diff
@@ -4239,6 +4239,18 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
 
+	// check whether the path's index has a tidb_shard() prefix and the index column count more than 1
+	// e.g. index(tidb_shard(a), a)
+	for _, path := range ds.possibleAccessPaths {
```
At L4178, It hasn't rewrite the virtual expression for virtual gc columns.
I placed my code at L4178 firstly,I can only judge whether it contain expression index but not shard index.
Maybe reomove the rewrite code ahead, but I can't evaluate the affect.

cfzjywxk replied:
I think the rewrite could be put above the path processing, we could do the refactor and this could be verified by the planner reviewer.

TonsnakeLin commented on util/ranger/ranger_test.go:
```diff
@@ -1433,3 +1436,431 @@ func TestPrefixIndexAppendPointRanges(t *testing.T) {
+	testKit.MustExec("create table testz(id int primary key clustered, a int, b int, " +
+		"unique key uk_expr((tidb_shard(a+b)),a))")
+
+	tests := []struct {
```
I will add it later

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+}
+
+// GetFieldsFromExpr get all fields from input expresion virtaulExpr
+func GetFieldsFromExpr(virtaulExpr *expression.ScalarFunction) []*expression.Column {
```
I will fix it later

TonsnakeLin replied:
ok

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+//     is `b` that's not the column in `tidb_shard(a)`.
+// @param  sf	"IN" function, e.g. `a IN (1, 2, 3)`
+func NeedAddColumn4InCond(cols []*expression.Column, accessCond []expression.Expression, sf *expression.ScalarFunction) bool {
+	if cols == nil || accessCond == nil || sf == nil {
```
I will fix it later

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+// @param[in] cols        the columns of shard index, such as [tidb_shard(a), a, ...]
+// @param[in] lengths     the length for every column of shard index
+// @retval - the new condition after adding tidb_shard() prefix
+func AddExpr4EqAndInCondition(sctx sessionctx.Context, conditions []expression.Expression,
```
Oh....., I think it's not necessary

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+			continue
+		}
+		// if the same field appear twice or more, don't add tidb_shard()
+		// e.g. `WHERE a > 100 and a < 200`
```
No, it add nothing to the condition

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+				return NeedAddColumn4EqCond(cols, accessCond, columnValues)
+			case ast.In:
+				return NeedAddColumn4InCond(cols, accessCond, f)
+			}
```
It return an error when some system error produced, such as memory allocate fail. 
if the sql isn't satisfied to add tidb_shard prefix, it's not necessary to make an error.

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,386 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+func AddExpr4EqAndInCondition(sctx sessionctx.Context, conditions []expression.Expression,
+	cols []*expression.Column) ([]expression.Expression, error) {
+
+	accesses := make([]expression.Expression, len(cols))
```
ok

TonsnakeLin commented:
/run unit-test

you06 commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

cfzjywxk commented:
@TonsnakeLin 
If the code change is finished you could trigger the "re-request" review button in the "reviwers" column, or add some a comment to ping the related reviewers.

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented on expression/builtin_miscellaneous.go:
```diff
@@ -92,6 +93,11 @@ var (
+)
+
+const (
+	tidbShardBucketCount = 256
```
@cfzjywxk PTAL again

TonsnakeLin commented on expression/column.go:
```diff
@@ -552,8 +553,8 @@ func ColInfo2Col(cols []*Column, col *model.ColumnInfo) *Column {
-// indexCol2Col finds the corresponding column of the IndexColumn in a column slice.
-func indexCol2Col(colInfos []*model.ColumnInfo, cols []*Column, col *model.IndexColumn) *Column {
+// IndexCol2Col finds the corresponding column of the IndexColumn in a column slice.
+func IndexCol2Col(colInfos []*model.ColumnInfo, cols []*Column, col *model.IndexColumn) *Column {
```
@cfzjywxk PTAL again

TonsnakeLin commented on planner/core/rule_predicate_push_down.go:
```diff
@@ -118,6 +129,9 @@ func (p *LogicalUnionScan) PredicatePushDown(predicates []expression.Expression)
 func (ds *DataSource) PredicatePushDown(predicates []expression.Expression) ([]expression.Expression, LogicalPlan) {
 	predicates = expression.PropagateConstant(ds.ctx, predicates)
 	predicates = DeleteTrueExprs(ds, predicates)
+	// Add tidb_shard() prefix to the condtion for shard index in some scenarios
```
@cfzjywxk PTAL again

TonsnakeLin commented on planner/core/rule_predicate_push_down_test.go:
```diff
@@ -0,0 +1,124 @@
+	c.Assert(s.store.Close(), IsNil)
+}
+
+func (s *testSharIndexResults) TestSharIndexPlan(c *C) {
```
@cfzjywxk PTAL again!  
Add explain test cases.

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+	}
+
+	// remove the accesses from newConditions
+	newConditions := make([]expression.Expression, 0, len(conditions))
```
@cfzjywxk PTAL again!
make a slice when it is necessary

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+	// add Gc condtion for accesses and return new condition to newAccesses
+	newAccesses, err := AddGcColumnCond(sctx, cols, accesses, columnValues)
+	if err != nil {
+		return conditions, err
```
@you06 @cfzjywxk PTAL again!
return the error if it was produced

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+	vi := &valueInfo{false, &evaluated}
+	con := &expression.Constant{Value: evaluated, RetType: cols[0].RetType}
+	// make a tidb_shard() function, e.g. `tidb_shard(a) = 8`
+	cond, err := expression.NewFunction(sctx, ast.EQ, cols[0].RetType, cols[0], con)
```
@you06 @cfzjywxk PTAL again!
I replace NewFunctionInternal to NewFunction that can return an error if something failed.

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+
+		// tmpArg1 is like `tidb_shard(a) = 8`, tmpArg1 is like `a = 100`
+		exprCon := &expression.Constant{Value: exprVal, RetType: cols[0].RetType}
+		tmpArg1, err1 := expression.NewFunction(sctx, ast.EQ, cols[0].RetType, cols[0], exprCon)
```
@cfzjywxk @you06 PTAL again!
All kinds of errors are returned.

TonsnakeLin commented on planner/core/rule_predicate_push_down.go:
```diff
@@ -638,3 +652,119 @@ func (p *LogicalMemTable) PredicatePushDown(predicates []expression.Expression)
+		}
+		newConds, err = ds.addExprPrefixCond(sc, path, newConds)
+		if err != nil {
+			logutil.BgLogger().Error("Add tidb_shard expression failed", zap.Error(err))
```
@you06 @cfzjywxk PTAL again!
If any errors produced, don't add tidb_shard expression and log it.

cfzjywxk replied:
The information may not be enough, we could print more about:
- connection id
- table name
- index name and column

dianqihanwangzi commented:
/run-check_dev_2

dianqihanwangzi commented:
/run-check_dev_2

youjiali1995 commented on expression/builtin.go:
```diff
@@ -781,6 +781,7 @@ var funcs = map[string]functionClass{
 	ast.VitessHash:      &vitessHashFunctionClass{baseFunctionClass{ast.VitessHash, 1, 1}},
 	ast.UUIDToBin:       &uuidToBinFunctionClass{baseFunctionClass{ast.UUIDToBin, 1, 2}},
 	ast.BinToUUID:       &binToUUIDFunctionClass{baseFunctionClass{ast.BinToUUID, 1, 2}},
+	ast.TidbShard:       &tidbShardFunctionClass{baseFunctionClass{ast.TidbShard, 1, 1}},
```
`TiDBShard` is more proper. https://github.com/golang/go/wiki/CodeReviewComments#initialisms

you06 commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+	return true
+}
+
+// ExtractColumnsFromExpr get all fields from input expresion virtaulExpr
```
```suggestion
// ExtractColumnsFromExpr get all fields from input expression virtaulExpr
```

you06 commented on util/ranger/ranger_test.go:
```diff
@@ -1765,3 +1768,431 @@ create table t(
+	// test ExtractColumnsFromExpr function
+	// -------------------------------------------
+	// input is nil
+	require.True(t, len(ranger.ExtractColumnsFromExpr(nil)) == 0)
```
```suggestion
	require.Equal(t, len(ranger.ExtractColumnsFromExpr(nil)), 0)
```
Use `require.Equal` instead, it offers more information when test failed.

you06 commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+		if sf, ok := arg.(*expression.ScalarFunction); ok {
+			fields = append(fields, ExtractColumnsFromExpr(sf)...)
+		} else if c, ok := arg.(*expression.Column); ok {
+			if !c.InColumnArray(fields) {
```
This ensures no duplicated columns between `fields` and the new column, however, if two fields from sub expressions have duplicated columns, they are merged into one in `fields = append(fields, ExtractColumnsFromExpr(sf)...)`, is this situation expected?

TonsnakeLin replied:
Yes, I want support index((tidb_shard(expr1(a),expr(a)),a, b) at first, but now this scenes is filted by IsValidShardIndex.

you06 commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+// IsValidShardIndex Check whether the definition of shard index is valid. The form of index
+// shoulde like `index(tidb_shard(a), a, ....)`.
+// 1) the column count shoudle be >= 2
+// 2) the first column shoudle be tidb_shard(xxx)
```
May search globally for "shoudle", and replace them to "should"

you06 reviewed (commented):
Rest lgtm, need some reviews from planner experts.

TonsnakeLin commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+	andType := types.NewFieldType(mysql.TypeTiny)
+
+	sf := accessesCond[1].(*expression.ScalarFunction)
+	c, _ := sf.GetArgs()[0].(*expression.Column)
```
Yes, it's more clean， but I can't find out  the code "?" in tidb-server, Almost  "x,_ " this way .
I want to add the check for transformation as you said before.

TonsnakeLin replied:
I'll fix it

you06 commented on util/ranger/ranger_test.go:
```diff
@@ -1765,3 +1768,431 @@ create table t(
+	exprLt := expression.NewFunctionInternal(sctx, ast.LT, longlongType, col4, con5)
+	andExpr2 := expression.NewFunctionInternal(sctx, ast.And, longlongType, exprLt, con5)
+	orExpr2 := expression.NewFunctionInternal(sctx, ast.Or, longlongType, andExpr1, andExpr2)
+	require.True(t, len(ranger.ExtractColumnsFromExpr(orExpr2.(*expression.ScalarFunction))) == 3)
```
These `require.True` can be replaced to `require.Equal` also.

wjhuang2016 commented on expression/builtin_miscellaneous.go:
```diff
@@ -1304,3 +1310,49 @@ func swapStringUUID(str string) string {
+
+	bf.tp.Flen = 4 //64 bit unsigned
+	bf.tp.Flag |= mysql.UnsignedFlag
+	types.SetBinChsClnFlag(bf.tp)
```
No need to set it.

TonsnakeLin replied:
I copy from `(c *vitessHashFunctionClass) getFunction`, and my function `tidb_shard(x)` equals `vitess_hash%256`,

wjhuang2016 commented on expression/builtin_miscellaneous.go:
```diff
@@ -1304,3 +1310,49 @@ func swapStringUUID(str string) string {
+		return nil, err
+	}
+
+	bf.tp.Flen = 4 //64 bit unsigned
```
Why set it to 4?

TonsnakeLin replied:
In fact, it's only 3 digits for value of tidb_shard(x) which equals vitess_hash(x)%256, I relaxed it to 4. 
Do you think it should be set 3？

wjhuang2016 commented on expression/builtin_miscellaneous.go:
```diff
@@ -92,6 +93,11 @@ var (
+)
+
+const (
+	tidbShardBucketCount = 256
```
What if we want to change it?

TonsnakeLin replied:
We don't change it, 256 buckets are enough

wjhuang2016 commented on expression/builtin_miscellaneous_test.go:
```diff
@@ -577,3 +577,37 @@ func TestBinToUUID(t *testing.T) {
+	fc := funcs[ast.TiDBShard]
+
+	// tidb_shar(-1) == 81, ......
+	args := makeDatums([]int{-1, 0, 1, 9999999999999999})
```
Need to cover some boundary values

TonsnakeLin replied:
got it

wjhuang2016 commented on planner/core/logical_plan_builder.go:
```diff
@@ -4226,6 +4226,18 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
+			col := expression.IndexCol2Col(ds.Columns, ds.schema.Columns, path.Index.Columns[0])
+			if col != nil && expression.GcColumnExprIsTidbShard(col.VirtualExpr) && len(path.Index.Columns) > 1 {
+				path.IsShardIndexPath = true
+				ds.containExprPrefixUk = true
```
Doesn't check the unique?

TonsnakeLin replied:
got it

TonsnakeLin replied:
In the `(ds *DataSource) AddPrefix4ShardIndexes` , use path.IsUkShardIndex() to check whether it is a unique shard index.
I wanted to support shard index for both non-unique and unique index at first, but we decided to support shard index only for unique index, I left it for future.
Do you think I should remove fix it?

TonsnakeLin replied:
I fixed it

wjhuang2016 commented on planner/core/rule_column_pruning.go:
```diff
@@ -309,7 +309,9 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column, opt *log
 	originColumns := ds.Columns
 	for i := len(used) - 1; i >= 0; i-- {
-		if !used[i] && !exprUsed[i] {
+		// if ds has a shard index, it can't prune the generated column of shard index
```
```suggestion
		// If ds has a shard index, it can't prune the generated column of shard index.
```

wjhuang2016 commented on planner/core/rule_predicate_push_down.go:
```diff
@@ -23,12 +23,23 @@ import (
 
+type exprPrefixAdder struct {
+	sctx      sessionctx.Context
+	OrigConds []expression.Expression
```
Need comments.

wjhuang2016 commented on planner/core/rule_predicate_push_down.go:
```diff
@@ -726,3 +740,119 @@ func appendAddSelectionTraceStep(p LogicalPlan, child LogicalPlan, sel *LogicalS
 	opt.appendStepToCurrent(sel.ID(), sel.TP(), reason, action)
 }
+
+// AddPrefix4ShardIndexes Add expression prefix for shard index. e.g. an index is test.uk(tidb_shard(a), a).
```
```suggestion
// AddPrefix4ShardIndexes adds expression prefix for shard index. e.g. an index is test.uk(tidb_shard(a), a).
```

wjhuang2016 commented on planner/core/rule_predicate_push_down_test.go:
```diff
@@ -0,0 +1,124 @@
+	dom   *domain.Domain
+}
+
+func (s *testSharIndexResults) SetUpTest(c *C) {
```
move them to explain_test.

wjhuang2016 commented on planner/util/path.go:
```diff
@@ -64,6 +64,9 @@ type AccessPath struct {
 	// IsSingleScan indicates whether the path is a single index/table scan or table access after index scan.
 	IsSingleScan bool
+
+	// maybe added in model.IndexInfo better, but the cache of model.IndexInfo may lead side effect
```
```suggestion
	// Maybe added in model.IndexInfo better, but the cache of model.IndexInfo may lead side effect.
```

wjhuang2016 commented on expression/builtin_miscellaneous.go:
```diff
@@ -1304,3 +1310,49 @@ func swapStringUUID(str string) string {
+}
+
+// evalInt evals tidb_shard(int64).
+func (b *builtinTidbShardSig) evalInt(row chunk.Row) (int64, bool, error) {
```
vecEvalInt version()?

TonsnakeLin replied:
Thanks, but there is no scenes to use vecEvalInt now

TonsnakeLin commented:
/run-check_dev

TonsnakeLin commented:
/run-check_dev_2

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

cfzjywxk commented on planner/core/logical_plan_builder.go:
```diff
@@ -4226,6 +4226,21 @@ func (b *PlanBuilder) buildDataSource(ctx context.Context, tn *ast.TableName, as
+		if !path.IsTablePath() {
+			col := expression.IndexCol2Col(ds.Columns, ds.schema.Columns, path.Index.Columns[0])
+			if col != nil &&
+				expression.GcColumnExprIsTidbShard(col.VirtualExpr) &&
```
Could we add a flag indicating if it's a shard index so we don't need to much check here, also it's easier to control the actual meaning of `shared index` type.

TonsnakeLin replied:
I agree with you fully. At first，I want to add a new syntax `SHARD INDEX` and add a flag for table meta so that it's easier to judge whether it has a shard index and whether the index is shard index. But as you know , we every body consider that it can cause unnecessary trouble for the fully realized shard index by @tiancaiamao , I give it up at last. If I add a flag by expression `tidb_shard` not by key word `SHARD`, I will get much information about this expression and do the check for it as now at ddll phase, I think it's not suitable . e.g. primary key is a meta flag for a index by key word 'PRIMARY'.

cfzjywxk commented:
@wjhuang2016 @qw4990 @tiancaiamao @tangenta 
PTAL thx~

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

winoros commented on planner/core/logical_plans.go:
```diff
@@ -586,6 +586,10 @@ type DataSource struct {
+
+	// contain unique index and the first field is tidb_shard(),
+	// such as (tidb_shard(a), a ...), the fields are more than 2
+	containExprPrefixUk bool
```
This should be the property of the meta data of the table instead of the logical plan?

TonsnakeLin replied:
Yes, I did it as you said at first. But we know that tiancaimao will resolve the hot index fully in the future, and our team think that I should do this as simple as possible, so I don't import a new syntax or change the process of ddl.
Please see the review comment above by cfzjywxk.

winoros commented on planner/core/rule_column_pruning.go:
```diff
@@ -309,7 +309,9 @@ func (ds *DataSource) PruneColumns(parentUsedCols []*expression.Column, opt *log
-		if !used[i] && !exprUsed[i] {
+		// If ds has a shard index, it can't prune the generated column of shard index
+		// so add the new condtion `!ds.containExprPrefixUk`
+		if !used[i] && !exprUsed[i] && !ds.containExprPrefixUk {
```
Still curious why we just prune no columns when there's the index.

TonsnakeLin replied:
When we use `uk(tidb_shard(a), a)` , it creates a generated column like `shardfield int Generated Always as (tidb_shard(a))` implicitly. There's a sql like `SELECT * FROM t WHERE a = 10`,  it doesn't use the column `shardfield` so that `shardfield` was pruned and we can't use it later in `ds.Columns`.

winoros replied:
It's not about the added virtual column.
When it's a normal column and it's not used, it will still not be pruned since the `containExprPrefixUk` is true.

TonsnakeLin replied:
Yes, I am lazy when I implement this function because of the principle of 'as simple as possible'. 
I fixed it, please check it , thank you!!

winoros commented on planner/core/rule_predicate_push_down.go:
```diff
@@ -726,3 +742,123 @@ func appendAddSelectionTraceStep(p LogicalPlan, child LogicalPlan, sel *LogicalS
+	}
+
+	var err error
+	newConds := conds
```
This assignment would not change the memory address.
If we want a new slice pointing to a new memory address, we need to `make([]Expression)` and `copy(src, dest)`.

TonsnakeLin replied:
I do it to avoid unnecessary memory allocation and copy if it doesn't need add a prefix condition that was checked at the function call `ds.addExprPrefixCond`. If it need add a prefix condition, I make a new slice and copy the old to it. The address of `conds` was not changed.

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+// @param[in] cols        the columns of shard index, such as [tidb_shard(a), a, ...]
+// @retval - if the shard index is valid return true, otherwise return false
+func IsValidShardIndex(cols []*expression.Column) bool {
+	// definition of index shoulde like the form: index(tidb_shard(a), a, ....)
```
```suggestion
	// definition of index should be like the form: index(tidb_shard(a), a, ....)
```

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+}
+
+// ExtractColumnsFromExpr get all fields from input expression virtaulExpr
+func ExtractColumnsFromExpr(virtaulExpr *expression.ScalarFunction) []*expression.Column {
```
There is `expression.ExtractColumns`.

TonsnakeLin replied:
I need to remove the duplicate columns in the `virtaulExpr`

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+}
+
+// IsEqOrInFunc check whetere it si equal or in function
+func IsEqOrInFunc(expr expression.Expression) bool {
```
Seems that I only find this expression in the testing.

TonsnakeLin replied:
I‘ll remove it

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+	matchedKeyFldCnt := 0
+
+	// the columns of shard index shoude be more than 2, like (tidb_shard(a),a,...)
+	if len(cols) < 2 ||
```
Why we don't check the `len(cols) < 2` in the function `NeedAddGcColumn4ShardIndex`?

TonsnakeLin replied:
I'll add `len(cols) < 2` to `NeedAddGcColumn4ShardIndex`

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+
+	// the columns of shard index shoude be more than 2, like (tidb_shard(a),a,...)
+	if len(cols) < 2 ||
+		len(accessCond) < 2 ||
```
And we've checked this in the function `NeedAddGcColumn4ShardIndex `.

TonsnakeLin replied:
Oh, I did it as you said at first. There are too many review comments, some people think it should check again at inner function to add the robustness though I think it was necessary.
I'll remove it.

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+//     is `b` that's not the column in `tidb_shard(a)`.
+// @param  sf	"IN" function, e.g. `a IN (1, 2, 3)`
+func NeedAddColumn4InCond(cols []*expression.Column, accessCond []expression.Expression, sf *expression.ScalarFunction) bool {
+	if len(cols) == 0 || len(accessCond) == 0 || sf == nil {
```
Also curious do we need this checking?

TonsnakeLin replied:
Oh!! It's still by review comments.

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+//     e.g. uk(tidb_shard(a), a). If the condtions is `WHERE b in (1, 2, 3)`, the first param of "IN" function
+//     is `b` that's not the column in `tidb_shard(a)`.
+// @param  sf	"IN" function, e.g. `a IN (1, 2, 3)`
+func NeedAddColumn4InCond(cols []*expression.Column, accessCond []expression.Expression, sf *expression.ScalarFunction) bool {
```
Need this function to be exported?

TonsnakeLin replied:
For unit test.

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+// It checks whether EQ condtions need to be added tidb_shard() prefix.
+// (1) columns in accessCond are all columns of the index except the first.
+// (2) every column in accessCond has a constan value
+func NeedAddColumn4EqCond(cols []*expression.Column,
```
ditto

TonsnakeLin replied:
For unit test

winoros commented on planner/core/rule_predicate_push_down.go:
```diff
@@ -726,3 +742,123 @@ func appendAddSelectionTraceStep(p LogicalPlan, child LogicalPlan, sel *LogicalS
+		if !path.IsUkShardIndexPath {
+			continue
+		}
+		newConds, err = ds.addExprPrefixCond(sc, path, newConds)
```
When we meet the error, do we want to directly return the old slice? If so, i think we can directly return nil slice when the `addExprPrefix4DNFCond` need to return an error.

TonsnakeLin replied:
Yes, it returns the old slice when we meet the error.
I think the two ways are both good, but the current doesn't need to check nil.

winoros commented on planner/core/rule_predicate_push_down.go:
```diff
@@ -726,3 +742,123 @@ func appendAddSelectionTraceStep(p LogicalPlan, child LogicalPlan, sel *LogicalS
+					return []expression.Expression{condition}, err
+				}
+				newAccessItems = append(newAccessItems, expression.ComposeCNFCondition(adder.sctx, accesses...))
+			} else if sf.FuncName.L == ast.EQ || sf.FuncName.L == ast.In {
```
The leaf of the DNF can still be a CNF(And clause), so we may need to call `addExprPrefix4CNFCond`.

TonsnakeLin replied:
Yes, I declared it in the doc https://github.com/pingcap/tidb/pull/31260/files#diff-5278acfa3ac95c90f2f08cb5ab1911b3892253a52e6d2fc5b172d42f47befce6R393
We scatter the hot index as simple as possible at first, It was asked to cover the certain scenes.

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+
+	sf := accessesCond[1].(*expression.ScalarFunction)
+	c := sf.GetArgs()[0].(*expression.Column)
+	var AndOrExpr expression.Expression
```
Capitalizing the first char of the local variable is rarely used.

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+
+		// tmpArg1 is like `tidb_shard(a) = 8`, tmpArg1 is like `a = 100`
+		exprCon := &expression.Constant{Value: exprVal, RetType: cols[0].RetType}
+		tmpArg1, err1 := expression.NewFunction(sctx, ast.EQ, cols[0].RetType, cols[0], exprCon)
```
The err can be directly reused.
We don't need to `err1`, `err2`, and `err3`. Just one `err` is okay.

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+// For param explanation, please refer to the function `AddGcColumnCond`.
+// @retval -  []expression.Expression   the new condtions after adding `tidb_shard() = xxx` prefix
+//            error                     if error gernerated, return error
+func AddGcColumn4InCond(sctx sessionctx.Context,
```
This func also doesn't need to be exported?

TonsnakeLin replied:
For unit test

winoros commented on util/ranger/detacher.go:
```diff
@@ -876,3 +877,402 @@ func MergeDNFItems4Col(ctx sessionctx.Context, dnfItems []expression.Expression)
+		return false
+	}
+
+	for _, cond := range accessCond[1:] {
```
Do we need to just make sure that only the first column is eq cond?
If the index is `(shard(a), a, b)`. `a = 1 and b > 2` can also be use to construct ranges.

TonsnakeLin replied:
We add the prefix just for point get.

winoros replied:
For index `shard(a), a, b`, The `a > 1` cannot be used but `a = 1 and b > 2` can.

TonsnakeLin replied:
we support shard index just for unique secondary index and point get plan. For index `(shard(a), a, b)`, the condition  `a = 1 and b > 2` can't make a point get plan.

TonsnakeLin commented:
/run-unit-tes

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-check_dev_2

TonsnakeLin commented:
/run-check_dev_2

TonsnakeLin commented:
/run-check_dev_2

winoros commented on util/ranger/ranger_test.go:
```diff
@@ -1774,3 +1777,418 @@ create table t(
+	// -------------------------------------------
+	// test NeedAddGcColumn4ShardIndex function
+	// -------------------------------------------
+	// ranger.valueInfo is not export by package, we can.t test NeedAddGcColumn4ShardIndex
```
There's some test in the `ranger` package.

TonsnakeLin commented:
/run-check_dev

TonsnakeLin commented:
/run-check_dev

TonsnakeLin commented on executor/explain_test.go:
```diff
@@ -386,3 +386,84 @@ func (s *testSuite) TestFix29401(c *C) {
 	tk.MustExec(" explain select /*+ inl_hash_join(t1) */ * from tt123 t1 join tt123 t2 on t1.b=t2.e;")
 }
+
+func (s *testSuite) TestShardIndexPlan(c *C) {
```
But I think it's not a bad thing here

wjhuang2016 replied:
We plan to migrate these SQL tests to explain test, so they will be removed finally.

TonsnakeLin replied:
done

TonsnakeLin commented:
/run-check_dev

TonsnakeLin commented:
> Any transaction-related test?

En, this feature  does not involve the flow of transaction control.

winoros reviewed (approved):
I noticed that this can be regarded as the implementation of the `HASH` type index for TIDB.
We can enlarge its scope later.

wjhuang2016 commented:
> > Any transaction-related test?
> 
> En, this feature does not involve the flow of transaction control.

In transaction, the plan is changed

TonsnakeLin commented:
> > > Any transaction-related test?
> > 
> > 
> > En, this feature does not involve the flow of transaction control.
> 
> In transaction, the plan is changed. We need to test unionScan

Optimizer implemented so many kinds of executors, But it's irrelevant for this PR. The core principle for this PR is that recognizing a shard index and adding `tidb_shard=xx` to the condition, It does not care about the specific corresponding execution plan because the plan is decided by optimizer.

Still, I added the test cases for [UnionScan plan that is executed in a tansaction](https://github.com/PingCAP-QE/shard-index-test/blob/master/functional_test.go#L662). I added some other cases for transaction too, like [this](https://github.com/PingCAP-QE/shard-index-test/blob/master/functional_test.go#L509).
@wjhuang2016 @cfzjywxk

TonsnakeLin commented:
/merge

youjiali1995 commented:
/merge

TonsnakeLin commented:
/assign @youjiali1995

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

TonsnakeLin commented:
/run-unit-test

youjiali1995 commented:
/merge

TonsnakeLin commented:
/run-unit-test

cfzjywxk commented:
/merge

youjiali1995 commented:
/run-unit-test

youjiali1995 commented:
Need one approval from @pingcap/tidb-configuration-reviewer
![image](https://user-images.githubusercontent.com/14819777/154636698-8c748f2b-d9d9-4065-9a26-7cb8c2158cc3.png)

cfzjywxk commented:
/merge

cfzjywxk commented:
/run-unit-test

TonsnakeLin commented:
re-run-mysql-test
